lib/ansible/modules/cloud/kubevirt/kubevirt_vm.py

```diff
@@ -249,92 +256,99 @@ def argspec(self):
+            if err:
+                self.fail_json(**err)
             else:
-                self._manage_state(False, resource_vm, existing, wait, wait_timeout)
-                return True
+                changed = True
+
+        if self.params.get('wait'):
+            k8s_obj = self._wait_for_state(new_state)
+
+        return changed, k8s_obj
 
     def execute_module(self):
         # Parse parameters specific for this module:
-        definition = virtdict()
         ephemeral = self.params.get('ephemeral')
         state = self.params.get('state')
-
-        if not ephemeral:
-            definition['spec']['running'] = state == 'running'
-
-        # Execute the CURD of VM:
-        template = definition['spec']['template']
-        kind = 'VirtualMachineInstance' if ephemeral else 'VirtualMachine'
-        dummy, definition = self.construct_vm_definition(kind, definition, template)
-        result = self.execute_crud(kind, definition)
-        changed = result['changed']
-
-        # Manage state of the VM:
-        if state in ['running', 'stopped']:
-            if not self.check_mode:
-                ret = self.manage_state(state)
-                changed = changed or ret
+        if ephemeral and state not in ('present', 'absent'):
+            raise TypeError("Ephemeral VMs only support 'present' and 'absent' states.")
+
+        _used_params = [name for name in self.params if self.params[name] is not None]
+        vm_spec_change = True if set(VM_SPEC_PARAMS).intersection(_used_params) else False
+        changed = False
+        method = ''
+
+        self.client = self.get_api_client()
+        self._vm_resource = self.find_supported_resource('VirtualMachine')
+        k8s_obj = self.get_resource(self._vm_resource)
+        if not self.check_mode and not vm_spec_change and state != 'absent' and not k8s_obj:
+            self.fail("It's impossible to create an empty VM or change state of a non-existent VM.")
+
+        # Changes in VM's spec or any changes to VMIs warrant a full CRUD, the latter because
+        # VMIs don't really have states to manage; they're either present or don't exist
+        # Also check_mode always warrants a CRUD, as that'll produce a sane result
+        if vm_spec_change or state in ['present', 'absent'] or ephemeral or self.check_mode:
+            definition = virtdict()
+            if not ephemeral:
+                definition['spec']['running'] = state == 'running'
+            template = definition['spec']['template']
+            kind = 'VirtualMachineInstance' if ephemeral else 'VirtualMachine'
+            dummy, definition = self.construct_vm_definition(kind, definition, template)
+            result = self.execute_crud(kind, definition)
+            changed = result['changed']
+            k8s_obj = result['result']
+            method = result['method']
+
+        if not ephemeral and state in ['running', 'stopped'] and not self.check_mode:
```

Comment:
So we don't wait for empheral VM? I think we should

Reply:
Also we should still call this in case of check mode, and say whether it will be changed

Reply:
1. Previous logic didn't wait for ephemeral VMIs either, so no functionality has been lost in that regard.
2. I'm not sure it's possible to wait for VMI shutdowns (with the current openshift-restclient lib), and having only partial support for ephemeral waits is not necessarily better than just having no support.
3. manage_state() call was also skipped when in check_mode, so this PR doesn't change anything there either.

In other words: while your suggestions can be attempted/considered, they are out of scope for this PR, as it was not my goal to add them here and they weren't present before, so no functionality is lost.

Reply:
1. It did. When I was last testing.
2. What's missing there? Just waiting for missing entity is ok, no?
3. That's right. But why not support it?

Reply:
1. In your code, `manage_state()` is called only `if state in ['running', 'stopped']:`, which aren't really valid/all states for VMIs. And the one thing I tried, failed[1]. So if there are scenarios where your wait code is working on VMIs, I was not able to find them. As far as I could tell, it never did work, which is why I didn't reimplement it in my code.
2. Yes, I'm not aware of how to do waiting for missing entity. The pvc module doesn't have that either. If a way was found, that would be a good thing to add both here for VMIs and in pvc.
3. I'm not saying we shouldn't, I'm just saying I'm not going to in this particular PR, as I have other stuff with higher priority and it wasn't my goal to implement that particular feature here.


[1] Current devel branch, vmi is Running:
```
[mmazur@klapek playbook]$ tail -n 3 vm-start.yml
        wait: yes
        ephemeral: yes
        state: running
[mmazur@klapek playbook]$ apb vm-start.yml
(…)
TASK [start] ************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "error": 422, "msg": "Failed to patch object: b'{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"admission webhook \\\\\"virtualmachineinstances-update-validator.kubevirt.io\\\\\" denied the request: spec.template in body is a forbidden property\",\"reason\":\"Invalid\",\"details\":{\"causes\":[{\"message\":\"spec.template in body is a forbidden property\"}]},\"code\":422}\\n'", "reason": "Unprocessable Entity", "status": 422}
```

Reply:
1. That's issue of creating whole ephemral VM: https://github.com/ansible/ansible/pull/54453 With this patch empheral VMs+waiting works ok again.

2. Waiting for 404 could work. I used that in other modules.

3. Sure.

Reply:
1. Ok, so your point is that it was intended to work with VMIs, it just had a bunch of issues. In that case, I'll extend the waiting to work with vmis as well.
2. Hehe, I'm a dumb dumb and didn't check the underlying code. KubernetesRawModule.perform_action() (which gets called within execute_crud()) is actually wait–aware and will wait for resource creation/deletion. With creation we need additional logic on top, since we're not interested in just the k8s object being created, but the VMI actually running (or being stopped in case of VM with status: stopped). However in the case of VMI deletion, one perform_action returns, that actually means the object was actually deleted and there's nothing else to wait for.

Bottom line: I'll add VMI running waiting and that will mean both ephemeral and non–ephemeral vms will have rock–solid wait support.